all: Improve and consolidate calculation of word count

DeinAlptraum opened the pull request:
Closes #10031

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/gohugoio/hugo?pullRequest=10032) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/gohugoio/hugo?pullRequest=10032) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/gohugoio/hugo?pullRequest=10032) it.</sub>

DeinAlptraum commented:
Yes, I believe it fits the description, including closing an open issue.

DeinAlptraum committed:
all: Improve and consolidate calculation of word count

Closes #10031

DeinAlptraum commented:
I have updated this branch to resolve conflicts, and have also improved the implementation slightly.

DeinAlptraum commented:
@jmooring no, there are no word separators in Chinese/Japanese. The concept of a "word" is not even clearly defined in those languages, which is why counting characters is the best one can do.
I just looked up Korean and it seems that they do use spaces. We could use a different formula here or only treat CJ differently if you prefer.

In either case, I don't think this affects the resolution of #10031: the problem is that, with `hasCJKLanguage` activated, any reading time is computed as if the text consists of _only_ CJK characters. As an example, a text that contains one CJK character and 10,000 non-CJK words currently results in a reading time of about 20mins (rather than the expected 47mins) because it uses the CJK formula, rather than "501 runes per minute" for the one CJK character and "213 words per minute" for the remaining 10,000 words
Splitting at spaces for CJK languages is not an issue even if they have spaces (like Korean) because we only count the number of runes, so this doesn't affect the result anyway.

This PR resolves that issue with mixed text, by computing separate counts for CJK characters as rune count and everything else via word count. This is still not perfect obviously (is Korean read at a similar speed as Chinese? what about Arabic anyway? etc.) but it's an improvement over the current situation at least, and eliminates an unnecessary setting.

DeinAlptraum commented:
@jmooring is there any more feedback or changes you'd like to see here?

jmooring commented:
I ran a quick performance comparison (before and after) on `strings.WordCount` on a page with approximated 2700 words, none of which contain CJK characters. The new implementation takes about 30% longer.

For this template function as well as `*cachedContentScope.contentPlain`, create some realistic benchmark tests (e.g., somewhere around 3000 words). Some ideas for optimization:

1. Bail out early if the string/content doesn't contain CJK
2. For non CJK, use `helpers.TotalWords` instead of converting the string to fields
3. My understanding is that integer division is faster that converting to float then dividing (look at reading time)

DeinAlptraum commented:
Thank you very much for your detailed feedback! I apologize for the delay, I was rather busy.

Regarding your feedback
#### Performance
I've adopted your optimized version. I also adapted it slightly: for CJK texts, it only looks at the first character of each word to determine whether it is a CJK word. This should be accurate enough for normal text and leads to a further speedup.

I've also added benchmarks for English text, mixed text and Chinese text, the results looked something like this (times in nanoseconds)
```
Current Master
    ASCII:      400
    Mixed:      170
    Chinese:    73

This PR (first version)
    ASCII:      520
    Mixed:      550
    Chinese:    73

Mooring Opt
    ASCII:      300
    Mixed:      340
    Chinese:    70

This PR (current version)
    ASCII:      300
    Mixed:      210
    Chinese:    67
```
To summarize, this is faster for pure CJK or English text, and slightly slower for mixed text.

Note that the "counting spaces instead of words" technique might lead to a count reduced by 1 (since usually `n` words are separated by `n-1` spaces).

I've also changed the reading time computation to go without in-between float64 conversion as you suggested. Note that this does now lead to a reading time of zero (due to integer division) when both the CJK count is below 501 and the non-CJK count is below 213. This necessitated a few more test adaptions.

#### Duplicate code
I eliminated the duplication in the places you mentioned, except for truncate.go: since it doesn't just count words but iterates through them one by one with custom behavior, I don't see the common function being usable here.
Besides, I've put the common word counting function into `common/hstrings/strings.go`; I'm not that familiar with the structure of the project, so I hope this is an acceptable place.

#### Other
I've adapted the PR title/description as you suggested.

Also:
> This resolves https://github.com/gohugoio/hugo/issues/10031 but it does not improve the CJK word count itself as, for example, a Chinese word may consist of one or more runes. So our CJK word count is too high in some (most?) cases.

As I mentioned before, "word" is not a clearly defined concept in Chinese (or Japanese for that matter, though I am not at all familiar with Korean). Counting characters is really the best you can do. In contrast to Western languages, reading speed or length limitations for documents etc. are, afaik, always defined via character counts.
The only problem I see is that our labeling as "word count" is then a bit confusing, but this is still an improvement over counting a space-less block of 1000 Chinese characters as a single word, which can happen on current master. And at least for the reading time computation, this is already a complete solution.

DeinAlptraum commented:
I've seen the CI failure at https://github.com/gohugoio/hugo/actions/runs/15906934442/job/44874052659?pr=10032
but it doesn't seem related to my change, or I don't know how to interpret this error.
I'll update the branch just in case.

jmooring commented:
I only spent a couple of minutes looking at this...

- Why remove the site config value? If `hasCJKLanguage` is `false` (the default) we can bypass detection. Think about a site with 1000 pages in 12 languages, and only one of the languages is CJK. That's a lot of unnecessary detection.
- I haven't thought this through, but it seems like deprecating the `isCJKLanguage` front matter field in favor of a `hasCJKLanguage` front matter field would be the right thing to do. For example, you could set the site level `hasCJKLanguage` to `false`, and set the `hasCJKLanguage` front matter field to `true` (override) where needed, either directly in front matter or via cascade. Again, a way to bypass the detection logic.
- Why did you change the reading time calculation and expected test results?
- Don't bother editing docs.yaml; it's generated from code.

jmooring commented:
> Why did you change the reading time calculation

Sorry, that question was unclear. Before we did this:

```go
(wordcount + 212) / 213
```

You changed it to this:

```go
wordcount / 213
```

The original one is correct. Assume there are 200 words. The original calc results in 1, yours results in zero. Integer division discards the remainder; there's no rounding.

> idea for a better name

How about just "hstrings.CountWords` ? That's what it does.

Also, cplease squash commits with a final commit message of 

```text
all: Improve and consolidate calculation of word count

Closes #10031
```

Then just force push changes thereafter.

I'll look at the rest of this laster.

Qodo-Free-For-OSS commented:
Hi, common/hstrings.CountWords classifies a token as CJK or non-CJK by inspecting only its first rune, so tokens like "a这是" or "Hugo日本" are counted as 1 non-CJK word and the embedded CJK runes are not counted. This causes incorrect WordCount/ReadingTime/Summary and tpl strings.CountWords results for mixed-language tokens.

**Severity:** action required | **Category:** correctness

**How to fix:** Count CJK runes within token

**Agent prompt to fix** - you can give this to your LLM of choice:

> ### Issue description
> `common/hstrings.CountWords` decides whether a token is CJK based only on the first rune. This undercounts tokens that start with non‑CJK but contain CJK runes (e.g. `a这是`), because those CJK runes never contribute to the returned counts.
> 
> ### Issue Context
> This function is now used by:
> - `hugolib` page `WordCount`/`ReadingTime`
> - HTML summary extraction (`ExtractSummaryFromHTML`)
> - template func `strings.CountWords`
> 
> ### Fix Focus Areas
> - common/hstrings/strings.go[133-159]
> 
> ### Implementation notes
> - While iterating each `word` from `strings.Fields`, scan the full token and count CJK runes inside it (or count runes as CJK/non-CJK segments), instead of relying on the first rune.
> - Ensure mixed tokens contribute both to `nNonCJK` (at least 1 for the non-CJK part, if desired) and `nCJK` for the CJK runes, or follow the previous behavior more directly (any token containing CJK counted by rune length). Add/adjust unit tests for mixed tokens like `a这是` and `Hugo日本`.

We noticed a couple of other issues in this PR as well - happy to share if helpful.

---
*Found by [Qodo](https://github.com/marketplace/qodo-merge-pro-for-open-source). Free code review for open-source maintainers.*

DeinAlptraum commented:
As good an opportunity as any: 
I'm closing this, not bothering to update this anymore. This PR has been open for 4 years now, and it's been a year since the last update. Resolving conflicts seems like it's just a waste of time.
Besides, I'm not talking to AI bots.
